File: app/src/main/java/com/akhbulatov/githubsample/ui/main/favoritesroot/favorites/FavoritesFragment.java
package com.akhbulatov.githubsample.ui.main.favoritesroot.favorites;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.akhbulatov.githubsample.App;
import com.akhbulatov.githubsample.R;
import com.akhbulatov.githubsample.models.UserDetails;

import java.util.List;

public final class FavoritesFragment extends Fragment implements FavoritesView {

    private RecyclerView favoritesRecyclerView;
    private ViewGroup loadingProgressLayout;
    private ViewGroup loadingErrorLayout;
    private TextView loadingErrorTextView;

    private FavoritesPresenter presenter;

    @Override public void onCreate(@Nullable Bundle savedInstanceState) {
        presenter = new FavoritesPresenter(App.getDataManager());
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState
    ) {
        return inflater.inflate(R.layout.fragment_favorites, container, false);
    }

    @Override public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViews(view);
        presenter.attachView(this);
    }

    @Override public void onDestroyView() {
        presenter.detachView();
        super.onDestroyView();
    }

    private void initViews(@NonNull View view) {
        Toolbar toolbar = view.findViewById(R.id.toolbar);
        toolbar.setTitle(R.string.favorites_title);
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white);
        toolbar.setNavigationOnClickListener(v -> presenter.onBackClicked());

        favoritesRecyclerView = view.findViewById(R.id.favoritesRecyclerView);
        favoritesRecyclerView.addItemDecoration(new DividerItemDecoration(
                favoritesRecyclerView.getContext(),
                DividerItemDecoration.VERTICAL)
        );

        loadingProgressLayout = view.findViewById(R.id.loadingProgressLayout);
        loadingErrorLayout = view.findViewById(R.id.loadingErrorLayout);
        loadingErrorTextView = view.findViewById(R.id.loadingErrorTextView);
    }

    @Override public void showFavorites(@NonNull List<UserDetails> favorites) {
        FavoritesAdapter adapter = new FavoritesAdapter(favorites);
        favoritesRecyclerView.setAdapter(adapter);
    }

    @Override public void showLoadingProgress(boolean show) {
        loadingProgressLayout.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @Override public void showEmptyFavorites() {
        loadingErrorTextView.setText(R.string.msg_empty_list);
        loadingErrorLayout.setVisibility(View.VISIBLE);
    }

    @Override public void backToUsers() {
        if (getActivity() != null) {
            getActivity().finish();
        }
    }
}

File: app/src/main/java/com/akhbulatov/githubsample/ui/main/favoritesroot/favorites/FavoritesFragment.kt
package com.akhbulatov.githubsample.ui.main.favoritesroot.favorites

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DividerItemDecoration
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.akhbulatov.githubsample.App
import com.akhbulatov.githubsample.R
import com.akhbulatov.githubsample.models.UserDetails
import kotlinx.android.synthetic.main.fragment_favorites.*
import kotlinx.android.synthetic.main.loading_error.*
import kotlinx.android.synthetic.main.loading_progress.*
import kotlinx.android.synthetic.main.toolbar.*

class FavoritesFragment : Fragment(), FavoritesView {

    private lateinit var presenter: FavoritesPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        presenter = FavoritesPresenter(App.dataManager)
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? = inflater.inflate(R.layout.fragment_favorites, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()
        presenter.attachView(this)
    }

    override fun onDestroyView() {
        presenter.detachView()
        super.onDestroyView()
    }

    private fun initViews() {
        toolbar.run {
            setTitle(R.string.favorites_title)
            setNavigationIcon(R.drawable.ic_arrow_back_white)
            setNavigationOnClickListener { presenter.onBackClicked() }
        }

        favoritesRecyclerView.addItemDecoration(
            DividerItemDecoration(favoritesRecyclerView.context, DividerItemDecoration.VERTICAL)
        )
    }

    override fun showFavorites(favorites: List<UserDetails>) {
        val adapter = FavoritesAdapter(favorites)
        favoritesRecyclerView.adapter = adapter
    }

    override fun showLoadingProgress(show: Boolean) {
        loadingProgressLayout.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun showEmptyFavorites() {
        loadingErrorTextView.setText(R.string.msg_empty_list)
        loadingErrorLayout.visibility = View.VISIBLE
    }

    override fun backToUsers() {
        activity?.finish()
    }
}